### ISSUE TITLE
Wrong error report of 'Use of moved value '

### ISSUE THREAD

**Issue by yipinliu** [intellij-rust/intellij-rust#7971]
## Environment

* **IntelliJ Rust plugin version:** 0.4.157.4174-212
* **Rust toolchain version:** 1.57.0-nightly (41dfaaa3c 2021-10-10) x86_64-pc-windows-msvc
* **IDE name and version:** CLion 2021.2.3 (CL-212.5457.51)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
The following code can be compiled successfully, but intellij-rust reports an error.
```rust,=
fn main() {
    let mut a = 1;
    let mut array = [&mut a];
    fun(array.iter_mut(), |v| *v += 1);
    println!("{}", a);
}

fn fun<'a, T, F>(iter: T, mut f: F)
where
    T: ExactSizeIterator<Item = &'a mut &'a mut i32>,
    F: FnMut(&mut i32)
{
    iter.for_each(|v| {
        for _ in 0..1 {
            f(v)
        }
    });
}
```
![image](https://user-images.githubusercontent.com/28223418/137486090-5eccd346-23cf-4966-bc94-a980d2dc15dd.png)

What is interesting is when I replace the `ExactSizeIterator` in line 10 with `Iterator`, it will work fine.
